doc/kernel/usermode/syscalls.rst

```diff
@@ -0,0 +1,259 @@
+.. _syscalls:
+
+System Calls
+############
+
+User threads run with a reduced set of privileges. Certain CPU instructions
+are forbidden to them, and they only have access to a limited part of the
+memory map. However, the kernel may allow user threads to perform operations
+that it would not normally be able to via system calls.
+
+This section describes how to declare new system calls and discusses a few
+implementation details relevant to them.
+
+Components
+==========
+
+All system calls have the following components:
+
+* A **C prototype** for the API, declared in some header under include/ and
+  prefixed with :c:macro:`__syscall`.  This prototype is never implemented
+  manually, instead it gets created by the ``scripts/gen_syscalls.py`` script.
+  What gets generated is an inline function which either calls the
+  implementation function directly (if called from supervisor mode) or goes
+  through privilege elevation and validation steps (if called from user
+  mode).
+
+* An **implementation function**, which is the real implementation of the
+  system call. The implementation function may assume that all parameters
+  passed in have been validated if it was invoked from user mode.
+
+* A **handler function**, which wraps the implementation function and does
+  validation of all the arguments passed in.
+
+C Prototype
+===========
+
+The C prototype represents how the API is invoked from either user or
+supervisor mode. For example, to initialize a semaphore:
+
+.. code-block:: c
+
+    __syscall void k_sem_init(struct k_sem *sem, unsigned int initial_count,
+                              unsigned int limit);
+
+The :c:macro:`__syscall` attribute is very special. To the C compiler, it
+simply expands to 'static inline'. However to the post-build
+``gen_syscalls.py`` script, it indicates that this API is a system call and
+generates the body of thefunction.  The gen_syscalls.py script does some
+parsing of the function prototype, to determine the data types of its return
+value and arguments, and has some limitations:
+
+* Array arguments must be passed in as pointers, not arrays. For example,
+  ``int foo[]`` or ``int foo[12]`` is not allowed, but should instead be
+  expressed as ``int *foo``.
+
+* Function pointers horribly confuse the limited parser. The workaround is
+  to typedef them first, and then express in the argument list in terms
+  of that typedef.
+
+* :c:macro:`__syscall` must be the first thing in the prototype and cannot
```

Comment:
delete "and cannot be preceded by anything else" (that's what "be the first thing" means)